cmd/admin-trace.go

```diff
@@ -378,11 +458,161 @@ func matchingOpts(ctx *cli.Context) (opts matchOpts) {
+		fatalIf(probe.NewError(e).Trace(ctx.String("rotate-trace-logs-size")), "Unable to parse input trace logs bytes.")
+	}
+
+	if ctx.String("rotate-trace-logs-frequency") != "" {
+		rotateTraceLogsFrequency, e = strconv.ParseUint(ctx.String("rotate-trace-logs-frequency"), 0, 64)
+		fatalIf(probe.NewError(e).Trace(ctx.String("rotate-trace-logs-frequency")), "Unable to parse input trace logs frequency.")
+	}
+
+	if ctx.String("rotate-trace-logs-pattern") != "" {
+		rotateTraceLogsPattern = ctx.String("rotate-trace-logs-pattern")
+		traceLogsPattern = rotateTraceLogsPattern
+	}
+
+	if ctx.String("rotate-trace-logs-compress") != "" {
+		rotateTraceLogsCompress = ctx.String("rotate-trace-logs-compress")
+		traceLogsCompress = rotateTraceLogsCompress
+	}
+
 	opts.requestSize = requestSize
 	opts.responseSize = responseSize
+	opts.rotateTraceLogsSize = rotateTraceLogsSize
+	opts.rotateTraceLogsFrequency = rotateTraceLogsFrequency
+	opts.rotateTraceLogsPattern = rotateTraceLogsPattern
+	opts.rotateTraceLogsCompress = rotateTraceLogsCompress
 	return
 }
 
+// Write trace log file
+func writeTracelogsFile(log string) {
+	traceLogsFileSize += len(log)
+	if _, err := traceLogsFile.WriteString(log + "\n"); err != nil {
+		traceLogsFile.Close()
+		fatalIf(probe.NewError(err), "Unable to write trace log file `"+defineTraceLogsFile()+"`")
+	}
+}
+
+// Rotate trace log file
+func rotateTraceLogsFile() {
+	defer func() {
+		// Close current file
+		if err := traceLogsFile.Close(); err != nil {
+			fatalIf(probe.NewError(err), "Unable to close current trace log file `"+traceLogsFilePath+"`")
+		}
+	}()
+	// Compress current file
+	if traceLogsCompress != "" {
+		// Remove raw file
+		defer func() {
+			if err := os.Remove(traceLogsFilePath); err != nil {
+				fatalIf(probe.NewError(err), "Unable to remove current trace log file `"+traceLogsFilePath+"`")
+			}
+		}()
+		switch traceLogsCompress {
+		case "gzip":
+			compressGzip()
+		case "zstd":
+			compressZstd()
+		default:
+			break
+		}
+	}
+}
+
+// gzip
+func compressGzip() {
+	compressFile, err := os.OpenFile(traceLogsFilePath+".gz", os.O_CREATE|os.O_RDWR, 0o644)
+	if err != nil {
+		compressFile.Close()
+		fatalIf(probe.NewError(err), "Unable to open trace log file archive for gzip `"+traceLogsFilePath+"`")
+	}
+	traceLogsFile.Seek(0, 0)
+	writer := gzip.NewWriter(compressFile)
+	_, err = io.Copy(writer, traceLogsFile)
+	if err != nil {
+		writer.Close()
+		compressFile.Close()
+		fatalIf(probe.NewError(err), "Unable to write trace log file archive for gzip `"+traceLogsFilePath+"`")
+	}
+	writer.Close()
+	compressFile.Close()
+}
+
+// zstandard
+func compressZstd() {
+	compressFile, err := os.OpenFile(traceLogsFilePath+".zst", os.O_CREATE|os.O_RDWR, 0o644)
+	if err != nil {
+		compressFile.Close()
+		fatalIf(probe.NewError(err), "Unable to open trace log file archive for zstd `"+traceLogsFilePath+"`")
+	}
+	traceLogsFile.Seek(0, 0)
```

Comment:
EDIT: Oh. You don't compress as you write, but you read back+compress when it cycles. I guess it is safer if you expect a crash, but it is a rather expensive.

I expected that you compressed directly to the output, so you didn't have to do a temporary file.

Reply:
@klauspost I also do this so that I can still pipe the live log output in case I want to see whats happening in realtime. Do you think avoiding the temporary file a better idea, all things considered?

Reply:
Im reviewing your "appending seamlessly" comment above.

Reply:
> Do you think avoiding the temporary file a better idea, all things considered?

It could easily be gigabytes. Recompressing that is not going to be instant.

The only downside is that we must be sure that we close both the compressor and the file before we exit for example on ctrl+c. Otherwise we end up with a broken file.

Reply:
To remove the compression bottleneck, I have set the compression to work asynchronously i.e. the logging will continue while compression is performed. I ensure that the closeable objects are closed. I also noticed that the best compression ratio for both gzip and zstd is "back compression" as you mention above.
Let me know your thoughts

Reply:
That is fine 👍🏼